File: examples/rust/examples/client_chat_message_stream.rs
//! ekoDB Rust Client - Chat Message Stream (SSE) Example
//!
//! Demonstrates streaming chat responses via HTTP SSE (Server-Sent Events).
//! This is a simpler alternative to WebSocket streaming.
//!
//! Run with: `cargo run --example client_chat_message_stream`

use ekodb_client::{ChatMessageRequest, Client, CreateChatSessionRequest};
use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    dotenv::dotenv().ok();

    println!("=== ekoDB Chat Message Stream (SSE) Example (Rust) ===\n");

    let client = Client::builder()
        .base_url(
            std::env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string()),
        )
        .api_key(std::env::var("API_BASE_KEY")?)
        .build()?;

    // Create a chat session
    let session = client
        .create_chat_session(
            CreateChatSessionRequest::new("openai").system_prompt("You are a helpful assistant."),
        )
        .await?;
    println!("Created session: {}", session.chat_id);

    // Stream a chat message via SSE
    println!("\nStreaming response for: 'What is ekoDB?'\n");
    let mut rx = client
        .chat_message_stream(&session.chat_id, ChatMessageRequest::new("What is ekoDB?"))
        .await?;

    use ekodb_client::websocket::ChatStreamEvent;
    while let Some(event) = rx.recv().await {
        match event {
            ChatStreamEvent::Chunk(text) => {
                print!("{}", text);
            }
            ChatStreamEvent::End {
                message_id,
                execution_time_ms,
                context_window,
                ..
            } => {
                println!("\n\n--- Stream complete ---");
                println!("Message ID: {}", message_id);
                println!("Execution time: {}ms", execution_time_ms);
                if let Some(cw) = context_window {
                    println!("Context window: {} tokens", cw);
                }
            }
            ChatStreamEvent::Error(err) => {
                eprintln!("\nError: {}", err);
            }
            _ => {}
        }
    }

    println!("\n✓ Chat message stream example completed");
    Ok(())
}

File: examples/kotlin/examples/ClientChatMessageStream.kt
// ekoDB Kotlin Client - Chat Message Stream (SSE) Example
//
// Demonstrates streaming chat responses via HTTP SSE (Server-Sent Events).

package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.ekodb.client.ChatStreamEvent
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.runBlocking

fun main(): Unit = runBlocking {
    println("=== ekoDB Chat Message Stream (SSE) Example (Kotlin) ===\n")

    val baseUrl = System.getenv("API_BASE_URL") ?: "http://localhost:8080"
    val apiKey = System.getenv("API_BASE_KEY")
        ?: throw IllegalStateException("API_BASE_KEY environment variable is required")

    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()

    var chatId: String? = null
    var failure: Throwable? = null
    try {
        // Create a chat session
        val session = client.createChatSession(buildJsonObject {
            putJsonArray("collections") {}
            put("llm_provider", "openai")
            put("llm_model", "gpt-4o-mini")
            put("system_prompt", "You are a helpful assistant.")
        })
        chatId = session["chat_id"]?.jsonPrimitive?.content
            ?: error("Chat session response did not include chat_id")
        println("Created session: $chatId")

        // Stream a chat message via SSE
        println("\nStreaming response for: 'What is ekoDB?'\n")
        val events = client.chatMessageStream(chatId, buildJsonObject {
            put("message", "What is ekoDB?")
        })

        var completed = false
        events.collect { event ->
            when (event) {
                is ChatStreamEvent.Chunk -> print(event.content)
                is ChatStreamEvent.End -> {
                    completed = true
                    println("\n\n--- Stream complete ---")
                    println("Message ID: ${event.messageId}")
                    println("Execution time: ${event.executionTimeMs}ms")
                    event.contextWindow?.let {
                        println("Context window: $it tokens")
                    }
                }
                is ChatStreamEvent.ToolCall -> println("[Tool] ${event.toolName}")
                is ChatStreamEvent.Error -> {
                    println("Error: ${event.error}")
                    error("Chat stream failed: ${event.error}")
                }
            }
        }
        check(completed) { "Chat stream ended before an end event" }
    } catch (error: Throwable) {
        failure = error
    } finally {
        try {
            chatId?.let { client.deleteChatSession(it) }
        } catch (cleanupError: Throwable) {
            val primary = failure
            if (primary == null) failure = cleanupError else primary.addSuppressed(cleanupError)
        }
        try {
            client.close()
        } catch (cleanupError: Throwable) {
            val primary = failure
            if (primary == null) failure = cleanupError else primary.addSuppressed(cleanupError)
        }
    }
    failure?.let { throw it }
    println("\n✓ Chat message stream example completed")
}
